Add downstream patches to fix testing in debian

lukpueh opened the pull request:
**Fixes issue #**:
None.


**Description of the changes being introduced by the pull request**:
A few tests call the python interpreter in a subprocess, e.g. to test our custom `process` module or to run simple in-toto step or inspection commands.

In debian, where we only build for python3, "python" is not available during build, and thus makes those tests fail.

This commit adds quilt downstream patches to change the used interpreter to "python3" in the relevant tests (this approach is backed by our Debian-man @h01ger).

Here's how I applied the patch and built in-toto 0.4.0 with tests:

```
 # Assuming chroot is set up
 # See https://wiki.debian.org/sbuild

 # Download upstream tarball
wget https://github.com/in-toto/in-toto/releases/download/v0.4.0/in-toto-0.4.0.tar.gz \
  -O python-in-toto_0.4.0.orig.tar.gz

 # Extract upstream tarball
tar xf python-in-toto_0.4.0.orig.tar.gz

 # Clone upstream repo for debian files and tests
git clone https://github.com/in-toto/in-toto.git in-toto-upstream

 # Copy debian files
cp -r in-toto-upstream/debian in-toto-0.4.0/

 # Change into unpacked tarball
cd in-toto-0.4.0

 # Apply patch in debian/patches/
quilt push

 # Build with tests
sbuild -A -s -d unstable
```


*An alternative for most tests might be, to use `sys.executable` instead of the hardcoded interpreter name. But this does not work for the inspection ("python ./scripts/tar ...") defined as string in "demo.layout.template".*

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Fix tests that call python interpreter for debian

A few tests call the python interpreter in a subprocess, e.g. to
test or custom `process` module or to run simple in-toto step or
inspection commands.

In debian, where we only build for python3, "python" is not
available during build, and thus makes those tests fail.

This commit adds quilt downstream patches to change the used
interpreter to "python3" in the relevant tests (this approach is
backed by our Debian-man @h01ger).

Here's how I applied the patches and built in-toto 0.4.0 with
tests:

```
 # Assuming chroot is set up
 # See https://wiki.debian.org/sbuild

 # Download upstream tarball
wget https://github.com/in-toto/in-toto/releases/download/v0.4.0/in-toto-0.4.0.tar.gz \
  -O python-in-toto_0.4.0.orig.tar.gz

 # Extract upstream tarball
tar xf python-in-toto_0.4.0.orig.tar.gz

 # Clone upstream repo for debian files and tests
git clone https://github.com/in-toto/in-toto.git in-toto-upstream

 # Copy debian files
cp -r in-toto-upstream/debian in-toto-0.4.0/

 # Change into unpacked tarball
cd in-toto-0.4.0

 # Apply patch in debian/patches/
quilt push

 # Build with tests
sbuild -A -s -d unstable
```

Note, an alternative for most tests would be, to use
`sys.executable` instead of the hardcoded interpreter name. But
this does not work for the inspection ("python ./scripts/tar ...")
defined as string in "demo.layout.template".

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Make debian upstream key minimal

See https://lintian.debian.org/tags/public-upstream-key-not-minimal.html

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Change section to "utils" in d/control

@h01ger suggests that this is a better fit than "devel"

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Add gpg to build dependencies in d/control

in-toto works well without gpg, but the tests, which are run on
build, don't. See
https://github.com/in-toto/in-toto/issues/264

So we add it as build dependency and keep it as runtime
recommendation.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Allow skip tests via DEB_BUILD_OPTIONS in d/rules

Running tests during build may be skipped by setting
`DEB_BUILD_OPTIONS=nocheck`.

See https://lintian.debian.org/tags/override_dh_auto_test-does-not-check-DEB_BUILD_OPTIONS.html

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Add required header fields to patch in d/patches

See https://lintian.debian.org/tags/quilt-patch-missing-description.html
and https://dep-team.pages.debian.net/deps/dep3/.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Release on debian as application not library

@h01ger claims that the executable trumps the library in terms of
naming, when a package has both, which in-toto does.

This commit changes "python[3]-in-toto" (library-style name) to just
"in-toto" (app-style name) in d/control and d/changelog.

See https://lintian.debian.org/tags/library-package-name-for-application.html

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Fix tab indentation in d/rules

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Fix gnupg dependency listing in d/control

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented:
So I've fixed all these problems (lintian warnings, missing gnupg dependency) and now tests keep failing at `test_in_toto_run_with_byproduct (tests.test_runlib.TestInTotoRun)`. This issues has been described in https://github.com/in-toto/in-toto/issues/282, but there's no fix for it yet. I'll look at it with a fresh pair of eyes early next week.

adityasaky commented:
Edit: Ah I get it now. Would you expect this to fail for in-toto regularly even if you're not specifically building it? I spun up a debian instance to see if tests would fail there when just cloning the repository and `test_in_toto_run_with_byproduct` did fail there with no STDOUT and STDERR entries that #285 should have got us. Is this the same thing you see? Or is this unrelated?

The link file I see is:

```
{
 "signatures": [],
 "signed": {
  "_type": "link",
  "byproducts": {
   "return-value": 0,
   "stderr": "",
   "stdout": ""
  },
  "command": [
   "python",
   "--version"
  ],
  "environment": {},
  "materials": {},
  "name": "test_step",
  "products": {}
 }
}
```

In `run_duplicate_streams`, the issue seems to be that the `stdout_reader` isn't reading the data from the tmpfile. I inspected the file itself - the output is there.

Commit:
Add downstream patch to hot-fix upstream bugs

The added patches fix two race conditions in in-toto's
process module.
They will be fixed upstream, but not in in-toto but in
securesystemslib, where the process module has been moved
to.

See https://github.com/secure-systems-lab/securesystemslib/pull/186/

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented:
Tests are running through when building downstream with the debian/* files in this branch now. :)
 See https://mentors.debian.net/package/in-toto for latest successful upload to mentors.

The commits in this PR do not only address @h01ger's review comments but also hot-fix some bugs from upstream (see #282 and https://github.com/secure-systems-lab/securesystemslib/pull/186).

SantiagoTorres commented on debian/upstream/signing-key.asc:
```diff
@@ -29,48 +24,29 @@ Vtvua2wFGj9DrxO+WQIPZrX/jJu7ntqxL2I4zLKRrvnMjAGtjT3I7zLsMvLaZN02
-=2tbg
------END PGP PUBLIC KEY BLOCK-----
\ No newline at end of file
+7L3JGDlPqKmG4qStrgzSxb/8Bpa5Ag0EVv7N3AEQALxccl4AppwwA4tCcxKdC0Ho
```
Just curious, why is your pubkey block shorter now?

lukpueh replied:
I exported with the `export-minimal` option ([as suggested by Lintian](https://lintian.debian.org/tags/public-upstream-key-not-minimal.html)). 

From the GPG manpages:

```
export-minimal
  Export the smallest key possible. This removes all signatures except the most
  recent self-signature on each user ID. This option is the  same as running
  the --edit-key command "minimize" before export except that the local copy of
  the key is not modified. Defaults to no.
```

lukpueh commented:
Updates were reviewed by @h01ger and used to put in-toto 0.4.0-1 into debian NEW queue.

Also see https://mentors.debian.net/package/in-toto
